### PR TITLE
feat: custom errors in 0.8.27

### PR DESCRIPTION
**Pull request by 8sunyuan**
Custom errors from 0.8.27 and `forge fmt src/contracts`

Do not merge right now. This will be merged into `dev` hand-in-hand with the slashing PR https://github.com/Layr-Labs/eigenlayer-contracts/pull/679. It is separated for the sake of review and less diff changes

Note: seems certora bash scripts are failing even after updating solc to 0.8.27

### CODE REVIEW

File: src/contracts/interfaces/IDelegationManager.sol [hunk 1]
```diff
@@ -15,6 +15,43 @@ import "./ISignatureUtils.sol";
...
+    /// @dev Thrown when provided `stakerOptOutWindowBlocks` cannot decrease.
+    error StakerOptOutWindowBlocksCannotDecrease();
+    /// @dev Thrown when provided `stakerOptOutWindowBlocks` exceeds maximum.
+    error StakerOptOutWindowBlocksExceedsMax();
+    /// @dev Thrown when attempting to withdraw before delay has elapsed.
+    error WithdrawalDelayNotElapsed();
+    /// @dev Thrown when provided delay exceeds maximum.
+    error WithdrawalDelayExeedsMax();
+    /// @dev Thrown when a withdraw amount larger than max is attempted.
+    error WithdrawalExeedsMax();
```
Comment (wadealexc) [inline src/contracts/interfaces/IDelegationManager.sol:49]: Spelling

File: src/contracts/interfaces/IDelegationManager.sol [hunk 2]
```diff
@@ -15,6 +15,43 @@ import "./ISignatureUtils.sol";
...
+    /// @dev Thrown when withdrawer is not the current caller.
+    error WithdrawerNotCaller();
```
Comment (wadealexc) [inline src/contracts/interfaces/IDelegationManager.sol:51]: I think we should collapse this, `CallerCannotUndelegate`, and `OnlyStrategyManagerOrEPM` into a single `UnauthorizedCaller` error

File: src/contracts/interfaces/IDelegationManager.sol [hunk 3]
```diff
@@ -15,6 +15,43 @@ import "./ISignatureUtils.sol";
...
+    /// @dev Thrown when attempting to use an expired eip-712 signature.
+    error SignatureExpired();
+    /// @dev Thrown when provided `stakerOptOutWindowBlocks` cannot decrease.
+    error StakerOptOutWindowBlocksCannotDecrease();
+    /// @dev Thrown when provided `stakerOptOutWindowBlocks` exceeds maximum.
+    error StakerOptOutWindowBlocksExceedsMax();
+    /// @dev Thrown when attempting to withdraw before delay has elapsed.
+    error WithdrawalDelayNotElapsed();
+    /// @dev Thrown when provided delay exceeds maximum.
+    error WithdrawalDelayExeedsMax();
```
Comment (wadealexc) [inline src/contracts/interfaces/IDelegationManager.sol:47]: Spelling

File: src/test/unit/DelegationUnit.t.sol [hunk 4]
```diff
@@ -2689,40 +2685,41 @@ contract DelegationManagerUnitTests_Undelegate is DelegationManagerUnitTests {
...
-        address invalidCaller
-    ) public filterFuzzedAddressInputs(invalidCaller) {
-        address staker = address(0x123);
-        // filter out addresses that are actually allowed to call the function
-        cheats.assume(invalidCaller != staker);
-        cheats.assume(invalidCaller != defaultOperator);
-        cheats.assume(invalidCaller != defaultApprover);
+    // function test_undelegate_revert_zeroAddress() public {
+    //     _registerOperatorWithBaseDetails(defaultOperator);
+    //     _delegateToOperatorWhoAcceptsAllStakers(address(0), defaultOperator);
```
Comment (wadealexc) [inline src/test/unit/DelegationUnit.t.sol:2695]: Why is this one commented out?

File: src/test/unit/DelegationUnit.t.sol [hunk 5]
```diff
@@ -2689,40 +2685,41 @@ contract DelegationManagerUnitTests_Undelegate is DelegationManagerUnitTests {
...
-        cheats.prank(invalidCaller);
-        cheats.expectRevert("DelegationManager.undelegate: caller cannot undelegate staker");
-        delegationManager.undelegate(staker);
-    }
+    //TODO: verify that this check is even needed
+    // /**
+    //  * @notice Verifies that the `undelegate` function has proper access controls (can only be called by the operator who the `staker` has delegated
+    //  * to or the operator's `delegationApprover`), or the staker themselves
+    //  */
+    // function testFuzz_undelegate_revert_invalidCaller(
```
Comment (wadealexc) [inline src/test/unit/DelegationUnit.t.sol:2707]: Why is this one commented out?

File: src/test/unit/DelegationUnit.t.sol (hunk 4 above)
Comment (0xClandestine) [reply src/test/unit/DelegationUnit.t.sol:2695]: It looks like the cancelSalt() no longer reverts if the provided salt has already been cancelled.

File: src/test/unit/AVSDirectoryUnit.t.sol [hunk 6]
```diff
@@ -310,40 +310,42 @@ contract AVSDirectoryUnitTests_operatorAVSRegisterationStatus is AVSDirectoryUni
...
+    // }
+
+    // NOTE: This check was removed.
+    // /// @notice Verifies that an operator cannot cancel the same salt twice
+    // function testFuzz_revert_whenCancellingSaltUsedToRegister(bytes32 salt) public {
+    //     address operator = cheats.addr(delegationSignerPrivateKey);
+    //     assertFalse(delegationManager.isOperator(operator), "bad test setup");
+    //     _registerOperatorWithBaseDetails(operator);
+
+    //     uint256 expiry = type(uint256).max;
```
Comment (wadealexc) [inline src/test/unit/AVSDirectoryUnit.t.sol:340]: Delete these

**Review by wadealexc — APPROVED** [review]
LGTM as soon as the cancelSalt tests are deleted (currently they are commented out)

Nice work!

**Comment by 0xClandestine** [comment]
Looks like `src/contracts/token/` was missed for custom errors. Will fix.

**Comment by wadealexc** [comment]
Closing in favor of just merging `slashing-magnitudes` directly into `dev`

**Pull request closed without merging**